src/hotspot/share/jfr/periodic/sampling/jfrCPUTimeThreadSampler.cpp

```diff
@@ -0,0 +1,827 @@
+  JavaThread* sampled_thread() const { return _sampled_thread; }
+
+  JfrAsyncStackTrace& stacktrace() { return _stacktrace; }
+
+  // Record a trace of the current thread
+  void record_trace(JavaThread* jt, void* ucontext) {
+    _stacktrace = JfrAsyncStackTrace(_frames, _max_frames);
+    set_sampled_thread(jt);
+    _type = NO_SAMPLE;
+    _error = 1;
+    _start_time = JfrTicks::now();
+    _end_time = JfrTicks::now();
+    if (!jt->in_deopt_handler() && !Universe::heap()->is_stw_gc_active())  {
+      ThreadInAsgct tia(jt);
+      if (thread_state_in_java(jt)) {
+        record_java_trace(jt, ucontext);
+      } else if (thread_state_in_native(jt)) {
+        record_native_trace(jt, ucontext);
+      }
+    }
+    _end_time = JfrTicks::now();
+  }
+
+private:
+
+  void record_java_trace(JavaThread* jt, void* ucontext) {
+    _type = JAVA_SAMPLE;
+    JfrGetCallTrace trace(true, jt);
+    frame topframe;
+    if (trace.get_topframe(ucontext, topframe)) {
+      _error = _stacktrace.record_async(jt, topframe) ? 0 : 11;
+    } else {
+      _error = 2;
+      return;
+     }
+  }
+
+  void record_native_trace(JavaThread* jt, void* ucontext) {
+    // When a thread is only attach it will be native without a last java frame
+   _type = NATIVE_SAMPLE;
+    _error = 1;
+    if (!jt->has_last_Java_frame()) {
+      _error = 3;
+      return;
+    }
+    frame topframe;
+    if (!jt->pd_get_top_frame_for_signal_handler(&topframe, ucontext, false)) {
+      _error = 6;
+      return;
+    }
+    frame first_java_frame;
+    Method* method = nullptr;
+    JfrGetCallTrace gct(false, jt);
+    if (!gct.find_top_frame(topframe, &method, first_java_frame)) {
+      _error = 4;
+      return;
+    }
+    if (method == nullptr) {
+      _error = 5;
+      return;
+    }
+    topframe = first_java_frame;
+    _error = _stacktrace.record_async(jt, topframe) ? 0 : 10;
+  }
+};
+
+// An atomic circular buffer of JfrTraces with a fixed size
+// Does not own any frames
+template <bool multiple_dequeuers, bool multipler_enqueuers>
+class JfrTraceQueue {
+  JfrCPUTimeTrace** _traces;
+  u4 _size;
+  volatile u4 _head;
+  volatile u4 _tail;
+
+public:
+  JfrTraceQueue(u4 size): _traces(JfrCHeapObj::new_array<JfrCPUTimeTrace*>(size)), _size(size), _head(0), _tail(0) {}
+
+  ~JfrTraceQueue() {
+    JfrCHeapObj::free(_traces, sizeof(JfrCPUTimeTrace) * _size);
+  }
+
+  bool is_empty() const { return _head == _tail; }
+  bool is_full() const { return (_head + 1) % _size == _tail; }
+
+  JfrCPUTimeTrace* dequeue() {
+    while (true) {
+      u4 current_tail = _tail;
+      u4 next_tail = (current_tail + 1) % _size;
+      if (current_tail == _tail) {
+        if (current_tail == _head) {
+          return nullptr; // queue is empty
+        }
+        if (Atomic::cmpxchg(&_tail, current_tail, next_tail, atomic_memory_order::memory_order_acquire) == current_tail) {
+          JfrCPUTimeTrace* trace = _traces[current_tail];
```

Comment:
Head could have been updated, but new trace not yet stored.
I'd suggest to spin-wait here until the trace is not empty, and then null out the slot once an element is extracted.